src/mlpack/methods/ann/loss_functions/cosine_embedding_loss.hpp

```diff
@@ -0,0 +1,156 @@
+   *                 to 0 and 1 respectively. Default value = 0.
+   */
+  CosineEmbeddingLoss(const double margin = 0.0, const bool takeMean = false);
+
+  /**
+   * Ordinary feed forward pass of a neural network.
+   *
+   * @param x1 Input data used for evaluating the given function.
+   * @param x2 Input data used for evaluating the given function.
+   * @param y Input data used to determine whether to calculate
+   *          cosine similarity or dissimilarity. It should only
+   *          contain values equal to 1 or -1.
+   */
+  template<
+    typename FirstTensor,
+    typename SecondTensor,
+    typename ThirdTensor
+  >
+  double Forward(const FirstTensor&& x1, const SecondTensor&& x2,
+      const ThirdTensor&& y);
+
+  /**
+   * Ordinary feed backward pass of a neural network. The negative log
+   * likelihood layer expects that the input contains log-probabilities for
+   * each class. The layer also expects a class index, in the range between 1
+   * and the number of classes, as target when calling the Forward function.
+   *
+   * @param x1 The propagated input activation.
+   * @param x2 The propagated input activation.
+   * @param y The target vector, that contains the class index in the range
+   *        between 1 and the number of classes.
+   * @param output The calculated error.
+   */
+  template<
+    typename FirstTensor,
+    typename SecondTensor,
+    typename ThirdTensor,
+    typename OutputTensor
+  >
+  void Backward(const FirstTensor&& x1,
+                const SecondTensor&& x2,
+                const ThirdTensor&& y,
+                const OutputTensor&& output);
+
+  //! Get the input parameter.
+  InputDataType& InputParameter() const { return inputParameter; }
+  //! Modify the input parameter.
+  InputDataType& InputParameter() { return inputParameter; }
+
+  //! Get the output parameter.
+  OutputDataType& OutputParameter() const { return outputParameter; }
+  //! Modify the output parameter.
+  OutputDataType& OutputParameter() { return outputParameter; }
+
+  //! Get the delta.
+  OutputDataType& Delta() const { return delta; }
+  //! Modify the delta.
+  OutputDataType& Delta() { return delta; }
+
+  //! Get the value of takeMean.
+  bool TakeMean() const { return takeMean; }
+  //! Modify the value of takeMean.
+  bool& TakeMean() { return takeMean; }
+
+  //! Get the value of margin.
+  bool Margin() const { return margin; }
+  //! Modify the value of takeMean.
+  bool& Margin() { return margin; }
+  
+
+  /**
+   * Serialize the layer.
+   */
+  template<typename Archive>
+  void serialize(Archive& /* ar */, const unsigned int /* version */);
+
+ private:
+  // Returns Cosine-Distance between two vectors.
+  template <typename FirstTensor, typename SecondTensor>
+  double CosineDistance(const FirstTensor&& x1,
+                        const SecondTensor&& x2)
+  {
+    return arma::dot(arma::normalise(x1), arma::normalise(x2));
```

Comment:
Didn't use norm_dot because it gives certain degree of error as it doesn't handle underflows and overflows very well.